game-app/game-core/src/main/java/games/strategy/triplea/ui/ResourceBar.java

```diff
@@ -53,51 +53,45 @@ private void setResources() {
 
   @Override
   public void gameDataChanged(final Change change) {
-    gameData.acquireReadLock();
-    try {
-      final GamePlayer player = gameData.getSequence().getStep().getPlayerId();
-      if (player != null) {
-        final IntegerMap<Resource> resourceIncomes =
-            AbstractEndTurnDelegate.findEstimatedIncome(player, gameData);
-        SwingUtilities.invokeLater(
-            () -> {
-              this.removeAll();
-              int count = 0;
-              for (final ResourceStat resourceStat : resourceStats) {
-                final Resource resource = resourceStat.resource;
-                if (!resource.isDisplayedFor(player)) {
-                  continue;
-                }
-                final double quantity =
-                    resourceStat.getValue(player, gameData, uiContext.getMapData());
-                final StringBuilder text =
-                    new StringBuilder(IStat.DECIMAL_FORMAT.format(quantity) + " (");
-                if (resourceIncomes.getInt(resource) >= 0) {
-                  text.append("+");
-                }
-                text.append(resourceIncomes.getInt(resource)).append(")");
-                final JLabel label =
-                    uiContext.getResourceImageFactory().getLabel(resource, text.toString());
-                label.setBorder(BorderFactory.createEmptyBorder(0, 10, 0, 10));
-                add(
-                    label,
-                    new GridBagConstraints(
-                        count++,
-                        0,
-                        1,
-                        1,
-                        0,
-                        1,
-                        GridBagConstraints.WEST,
-                        GridBagConstraints.BOTH,
-                        new Insets(0, 0, 0, 0),
-                        0,
-                        0));
-              }
-            });
-      }
-    } finally {
-      gameData.releaseReadLock();
+    // When there are multiple gameDataChanged() notifications in a row, for example
+    // from an "undo all" action, no need to do this repeatedly. This will just run
+    // once if the swing code runs after all the notifications have been received.
+    if (updateScheduled) {
+      return;
     }
+    updateScheduled = true;
+    SwingUtilities.invokeLater(
+        () -> {
+          updateScheduled = false;
+          final GamePlayer player;
+          final IntegerMap<Resource> resourceIncomes;
+          try {
+            gameData.acquireReadLock();
```

Comment:
This is bad practice. The event dispatch thread (swing thread) should never have to acquire a lock. If this lock is unavailable for some reason this will lead to a freezing UI and a bad user experience.

I'm actually not sure what's the best solution for this. In general your approach is fine. It's just this particular case where it's not optimal. Actually I think it would be better execute later invocations instead of the first one, so that the last change is actually picked up properly (in case it changes something). So that being said I have something like this in mind:
```java
private final AtomicInteger plannedUpdates = new AtomicInteger();

@Override
public void gameDataChanged(final Change change) {
  plannedUpdates.incrementAndGet();
  SwingUtilities.invokeLater(() -> {
    if (plannedUpdates.decrementAndGet() > 0) {
      return;
    }
    ThreadRunner.runInNewThread(() -> {
      // blocking code
      SwingUtilities.invokeLater(() -> {
        // ui code
      });
    });
  });
}
```

Obviously this seems unecessarily complicated because in the end we're switching to the EDT for the timing only, then to a new thread then back to the EDT. So maybe a solution like this would be better:
```java
private final AtomicInteger plannedUpdates = new AtomicInteger();

@Override
public void gameDataChanged(final Change change) {
  plannedUpdates.incrementAndGet();
  CompletableFuture.runAsync(() -> {
    if (plannedUpdates.decrementAndGet() > 0) {
      return;
    }
    // blocking code with error handling (otherwise any exception is just ignored)
    // or have a look at CompletableFutureUtils#logExceptionWhenComplete
    SwingUtilities.invokeLater(() -> {
      // ui code
    });
  });
}
```

This seems a little bit better, but honestly it seems questionable if the "plannedUpdates" counter is still as effective reducing the amount of calls because CompleteableFuture's Thread pool is most likely less busy than the EDT.

So yeah not really happy with this solution either. It definitely works, but the code is not really elegant.

The ideal solution would be to rethink the interface that is used for this because obviously it doesn't fit this use-case, but I'm pretty sure this wouldn't work without breaking network compatibility.

Reply:
> Actually I think it would be better execute later invocations instead of the first one, so that the last change is actually picked up properly (in case it changes something). So that being said I have something like this in mind:

Just to be clear, if a later invocation runs *after* the swing update, it will schedule another swing update, via the boolean. So there's no issue that the update wouldn't happen.

> So that being said I have something like this in mind:
> 
> ```java
> private final AtomicInteger plannedUpdates = new AtomicInteger();
> 
> @Override
> public void gameDataChanged(final Change change) {
>   plannedUpdates.incrementAndGet();
>   SwingUtilities.invokeLater(() -> {
>     if (plannedUpdates.decrementAndGet() > 0) {
>       return;
>     }
>     ThreadRunner.runInNewThread(() -> {
>       // blocking code
>       SwingUtilities.invokeLater(() -> {
>         // ui code
>       });
>     });
>   });
> }

I don't think we need a counter to increment or the three levels of tasks, we could just do two levels, something like:

> ```java
> @Override
> public void gameDataChanged(final Change change) {
>   if (updateScheduled) { return; }
>   updateScheduled = true;
>   ThreadRunner.runInNewThread(() -> {
>      updateScheduled = false;
>      // grab lock and get info from game state
>      SwingUtilities.invokeLater(() -> {
>        // ui code
>      });
>   } 
> ```
 
This way, we avoid doing the locking operation on the UI thread and we don't schedule the op if there is one scheduled that hasn't run. I think the main downside of the above is the extra overhead of starting a new thread each time.

Reply:
Looks like we can do: `BackgroundTaskRunner.runInBackground()` instead of always starting a new thread.

Reply:
Nevermind, it looks like `BackgroundTaskRunner` doesn't do the right thing, it's actually something that shows UI.

Reply:
I guess AsyncRunner.runAsync() does the job.

Reply:
Done, please take another look.

Reply:
A drive-by comment here, apologies for not being fully in-context:
- there are a number of ways to solving multi-threaded programming problems:
   - make everything single threaded (the EDT is an example of this)
   - use locks / semaphores
   - use thread-safe data

All immutable data is thread safe. Hence, ideally the UI never has a reference to any game data objects and only has references to immutable data. Essentially, the UI should be told to render X, and it should then render X. The references to GameData that the UI has is a mixture of logic in the UI layer, and with swing that gets kinda bad.. 

So.. I do discourage us using the 'gameData.acquireLock' mechanism and instead favor simply copying data and then using immutable data copies.